chunk/chunk_store.go

```diff
@@ -243,49 +243,104 @@ outer:
 	return filteredChunks, nil
 }
 
-func (c *Store) lookupMatchers(ctx context.Context, from, through model.Time, matchers []*metric.LabelMatcher) ([]Chunk, error) {
-	metricName, matchers, err := util.ExtractMetricNameFromMatchers(matchers)
+func (c *Store) lookupChunksByMatchers(ctx context.Context, from, through model.Time, matchers []*metric.LabelMatcher) ([]Chunk, error) {
+	userID, err := user.Extract(ctx)
 	if err != nil {
 		return nil, err
 	}
 
-	userID, err := user.Extract(ctx)
-	if err != nil {
-		return nil, err
+	// Get metric name from matchers
+	metricName, matchers, ok := util.ExtractMetricNameFromMatchers(matchers)
+
+	// If there is a metric name, return chunks using metric name
+	if ok {
+		return c.lookupChunksByMetricName(ctx, from, through, matchers, userID, metricName)
```

Comment:
My preference is to extract the userID from the context at the leaf of the function-tree, as we have to pass contexts for the sake of tracing.